File: test-suite/python/security/path_traversal_clean.py
from pathlib import Path

from flask import request, send_file
from werkzeug.utils import secure_filename

UPLOAD_ROOT = Path("/srv/app/uploads").resolve()


def validate_path(raw_name: str | None) -> Path:
    if not raw_name:
        raise ValueError("missing file")
    candidate = (UPLOAD_ROOT / raw_name).resolve()
    candidate.relative_to(UPLOAD_ROOT)
    return candidate


def flask_download():
    name = request.args.get("file")
    return send_file(validate_path(name))


def sanitized_filename_download():
    name = secure_filename(request.values.get("name", ""))
    target = UPLOAD_ROOT / name
    return target.read_bytes()


def sanitized_upload_save():
    uploaded = request.files["avatar"]
    name = secure_filename(uploaded.filename)
    target = UPLOAD_ROOT / name
    uploaded.save(target)
    return {"saved": name}


def containment_checked_upload_save():
    uploaded = request.files["document"]
    target = validate_path(uploaded.filename)
    uploaded.save(target)
    return {"saved": target.name}


def sanitized_upload_save_keyword():
    uploaded = request.files["invoice"]
    name = secure_filename(uploaded.filename)
    uploaded.save(dst=UPLOAD_ROOT / name)
    return {"saved": name}


def containment_checked_download():
    target = validate_path(request.args.get("document"))
    return target.read_bytes()


def flask_header_download():
    name = request.headers.get("X-File-Path")
    return send_file(validate_path(name))


def raw_header_open():
    target = validate_path(request.headers["X-Report-Path"])
    return target.read_bytes()


def django_meta_download(request):
    name = request.META.get("HTTP_X_FILE_PATH")
    return validate_path(name).read_bytes()

File: test-suite/js/security/path-traversal-clean.ts
import fs from "node:fs";
import path from "node:path";

type RequestLike = {
  query: Record<string, string | undefined>;
  params: Record<string, string | undefined>;
  body: Record<string, string | undefined>;
  get(name: string): string | undefined;
  header(name: string): string | undefined;
  file?: UploadedFile;
};

type ResponseLike = {
  sendFile(filePath: string): void;
  download(filePath: string): void;
};

type UploadedFile = {
  name: string;
  originalname: string;
  mv(destination: string): Promise<void>;
};

type HeaderBag = {
  get(name: string): string | null;
};

declare function headers(): Promise<HeaderBag>;

const UPLOAD_ROOT = path.resolve("/srv/app/uploads");

function validatePath(rawName: string | undefined): string {
  if (!rawName) {
    throw new Error("missing file name");
  }
  const target = path.resolve(UPLOAD_ROOT, rawName);
  const relative = path.relative(UPLOAD_ROOT, target);
  if (relative.startsWith("..") || path.isAbsolute(relative)) {
    throw new Error("path escaped upload root");
  }
  return target;
}

function sanitizeFilename(rawName: string | undefined): string {
  return path.basename(rawName ?? "upload.bin");
}

export function expressDownload(req: RequestLike, res: ResponseLike): void {
  const requested = req.query.file;
  res.sendFile(validatePath(requested));
}

export function rawRead(req: RequestLike): Buffer {
  const filePath = validatePath(req.params.name);
  return fs.readFileSync(filePath);
}

export async function uploadedFileSave(req: RequestLike): Promise<void> {
  const uploaded = req.file;
  if (!uploaded) {
    throw new Error("missing upload");
  }
  const target = path.join(UPLOAD_ROOT, sanitizeFilename(uploaded.originalname));
  await uploaded.mv(target);
}

export function bodyPathWrite(req: RequestLike): void {
  const output = validatePath(req.body.destination);
  fs.writeFileSync(output, "generated report");
}

export function directDownload(req: RequestLike, res: ResponseLike): void {
  res.download(validatePath(req.query.archive));
}

export function headerPathRead(req: RequestLike): Buffer {
  const requested = req.get("x-file-path");
  return fs.readFileSync(validatePath(requested));
}

export function headerPathWrite(req: RequestLike): void {
  const destination = validatePath(req.header("x-output-path"));
  fs.writeFileSync(destination, "generated report");
}

export async function nextHeaderPathRead(): Promise<Buffer> {
  const incomingHeaders = await headers();
  const requested = incomingHeaders.get("x-file-path");
  return fs.readFileSync(validatePath(requested ?? undefined));
}
